azurerm_role_definition unexpected EOF

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform
- Terraform version: 0.11.13 
- terraform-provider-azurerm_v1.30.1_x4
- terraform-provider-random_v2.1.2_x4 

### Affected Resource(s)

<!--- Please list the affected resources and data sources. --->

* `azurerm_role_definition`

### Terraform Configuration Files

```hcl
data "local_file" "devops_allow" {
  filename = "${path.module}/files/devops_allow.txt"
}

locals {
  devops_allow      = "${sort(split(",", trimspace(replace(replace(data.local_file.devops_allow.content,"\n",""), "\"", ""))))}"
}

resource "random_uuid" "devops" {}

resource "azurerm_role_definition" "devops" {
  name               = "iam-devops"
  role_definition_id = "${random_uuid.devops.result}"
  scope              = "${data.azurerm_subscription.primary.id}"

  description = "This is a sample custom rule set"

  permissions {
    actions          = ["${local.devops_allow}"]
    not_actions      = ["${local.devops_deny}"]
    data_actions     = ["${local.devops_data_allow}"]
    not_data_actions = ["${local.devops_data_deny}"]
  }

  assignable_scopes = ["${var.resource_group_id}"] # will be resource group id
}

```

### Debug Output

### Panic Output
https://gist.github.com/salimkapadia/5998cab4a6d16217bf4c10f067aaa8a4

### Context
I have moved all my permissions to a text file. These permissions look like this:
```
"*/read",
"Microsoft.Authorization/*/read",
"Microsoft.Cache/checknameavailability/action",
"Microsoft.Cache/locations/operationresults/read",
"Microsoft.Cache/operations/read",
```
 - The file has `LF` line endings. When the file is read, it contains new line characters which i replace with an empty string.
 - I also replace all double quotes with an empty string otherwise when its read, it tries to escape the quote. during the apply phase.

I have hit this error in TF v0.11.14, and when I switch to TF 0.12.8, I get the following error:

```
Error: rpc error: code = Unavailable desc = transport is closing
```

It looks like it could be related.

There is a TF dump
```
2019-09-06T15:40:06.429+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: [DEBUG] Determining which Resource Providers require Registration
2019-09-06T15:40:06.429+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: [DEBUG] All required Resource Providers are registered
2019/09/06 15:40:06 [TRACE] [walkApply] Exiting eval tree: provider.azurerm
2019/09/06 15:40:06 [TRACE] vertex "provider.azurerm": visit complete
2019/09/06 15:40:06 [TRACE] dag/walk: visiting "data.azurerm_subscriptions.available (prepare state)"
2019/09/06 15:40:06 [TRACE] vertex "data.azurerm_subscriptions.available (prepare state)": starting visit (*terraform.NodeApplyableResource)
2019/09/06 15:40:06 [TRACE] vertex "data.azurerm_subscriptions.available (prepare state)": evaluating
2019/09/06 15:40:06 [TRACE] [walkApply] Entering eval tree: data.azurerm_subscriptions.available (prepare state)
2019/09/06 15:40:06 [TRACE] <root>: eval: *terraform.EvalWriteResourceState
2019/09/06 15:40:06 [TRACE] dag/walk: visiting "module.roleAssignment.azurerm_role_assignment.roleAssignment (prepare state)"
2019/09/06 15:40:06 [TRACE] dag/walk: visiting "module.roleDefinition.azurerm_role_definition.roleDefinition (prepare state)"
2019/09/06 15:40:06 [TRACE] vertex "module.roleAssignment.azurerm_role_assignment.roleAssignment (prepare state)": starting visit (*terraform.NodeApplyableResource)
2019/09/06 15:40:06 [TRACE] vertex "module.roleDefinition.azurerm_role_definition.roleDefinition (prepare state)": starting visit (*terraform.NodeApplyableResource)
2019/09/06 15:40:06 [TRACE] vertex "module.roleAssignment.azurerm_role_assignment.roleAssignment (prepare state)": evaluating
2019/09/06 15:40:06 [TRACE] [walkApply] Entering eval tree: module.roleAssignment.azurerm_role_assignment.roleAssignment (prepare state)
2019/09/06 15:40:06 [TRACE] module.roleAssignment: eval: *terraform.EvalWriteResourceState
2019/09/06 15:40:06 [TRACE] vertex "module.roleDefinition.azurerm_role_definition.roleDefinition (prepare state)": evaluating
2019/09/06 15:40:06 [TRACE] [walkApply] Entering eval tree: module.roleDefinition.azurerm_role_definition.roleDefinition (prepare state)
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalWriteResourceState
2019/09/06 15:40:06 [TRACE] [walkApply] Exiting eval tree: data.azurerm_subscriptions.available (prepare state)
2019/09/06 15:40:06 [TRACE] vertex "data.azurerm_subscriptions.available (prepare state)": visit complete
2019/09/06 15:40:06 [TRACE] dag/walk: visiting "data.template_file.available_subscriptions (prepare state)"
2019/09/06 15:40:06 [TRACE] vertex "data.template_file.available_subscriptions (prepare state)": starting visit (*terraform.NodeApplyableResource)
2019/09/06 15:40:06 [TRACE] vertex "data.template_file.available_subscriptions (prepare state)": evaluating
2019/09/06 15:40:06 [TRACE] [walkApply] Entering eval tree: data.template_file.available_subscriptions (prepare state)
2019/09/06 15:40:06 [TRACE] <root>: eval: *terraform.EvalWriteResourceState
2019/09/06 15:40:06 [TRACE] GetResourceInstance: data.azurerm_subscriptions.available is a single instance
2019/09/06 15:40:06 [TRACE] [walkApply] Exiting eval tree: module.roleDefinition.azurerm_role_definition.roleDefinition (prepare state)
2019/09/06 15:40:06 [TRACE] vertex "module.roleDefinition.azurerm_role_definition.roleDefinition (prepare state)": visit complete
2019/09/06 15:40:06 [TRACE] [walkApply] Exiting eval tree: module.roleAssignment.azurerm_role_assignment.roleAssignment (prepare state)
2019/09/06 15:40:06 [TRACE] vertex "module.roleAssignment.azurerm_role_assignment.roleAssignment (prepare state)": visit complete
2019/09/06 15:40:06 [TRACE] [walkApply] Exiting eval tree: data.template_file.available_subscriptions (prepare state)
2019/09/06 15:40:06 [TRACE] vertex "data.template_file.available_subscriptions (prepare state)": visit complete
2019/09/06 15:40:06 [TRACE] dag/walk: visiting "provider.template (close)"
2019/09/06 15:40:06 [TRACE] vertex "provider.template (close)": starting visit (*terraform.graphNodeCloseProvider)
2019/09/06 15:40:06 [TRACE] vertex "provider.template (close)": evaluating
2019/09/06 15:40:06 [TRACE] [walkApply] Entering eval tree: provider.template (close)
2019/09/06 15:40:06 [TRACE] dag/walk: visiting "module.roleDefinition.var.assignableScopes"
2019/09/06 15:40:06 [TRACE] vertex "module.roleDefinition.var.assignableScopes": starting visit (*terraform.NodeApplyableModuleVariable)
2019/09/06 15:40:06 [TRACE] <root>: eval: *terraform.EvalCloseProvider
2019/09/06 15:40:06 [TRACE] GRPCProvider: Close
2019/09/06 15:40:06 [TRACE] vertex "module.roleDefinition.var.assignableScopes": evaluating
2019/09/06 15:40:06 [TRACE] [walkApply] Entering eval tree: module.roleDefinition.var.assignableScopes
2019/09/06 15:40:06 [TRACE] <root>: eval: *terraform.EvalSequence
2019/09/06 15:40:06 [TRACE] <root>: eval: *terraform.EvalOpFilter
2019/09/06 15:40:06 [TRACE] <root>: eval: *terraform.EvalModuleCallArgument
2019/09/06 15:40:06 [TRACE] GetResourceInstance: data.template_file.available_subscriptions has multiple keyed instances
2019-09-06T15:40:06.439+0100 [DEBUG] plugin: plugin process exited: path=/Users/Lee/.terraform.d/plugins/terraform-provider-template_v2.1.2_x4 pid=38360
2019-09-06T15:40:06.439+0100 [DEBUG] plugin: plugin exited
2019/09/06 15:40:06 [TRACE] [walkApply] Exiting eval tree: provider.template (close)
2019/09/06 15:40:06 [TRACE] vertex "provider.template (close)": visit complete
2019/09/06 15:40:06 [TRACE] <root>: eval: *terraform.EvalSetModuleCallArguments
2019/09/06 15:40:06 [TRACE] [walkApply] Exiting eval tree: module.roleDefinition.var.assignableScopes
2019/09/06 15:40:06 [TRACE] vertex "module.roleDefinition.var.assignableScopes": visit complete
2019/09/06 15:40:06 [TRACE] dag/walk: visiting "module.roleDefinition.azurerm_role_definition.roleDefinition[0]"
2019/09/06 15:40:06 [TRACE] vertex "module.roleDefinition.azurerm_role_definition.roleDefinition[0]": starting visit (*terraform.NodeApplyableResourceInstance)
2019/09/06 15:40:06 [TRACE] vertex "module.roleDefinition.azurerm_role_definition.roleDefinition[0]": evaluating
2019/09/06 15:40:06 [TRACE] [walkApply] Entering eval tree: module.roleDefinition.azurerm_role_definition.roleDefinition[0]
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalSequence
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalGetProvider
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalReadDiff
2019/09/06 15:40:06 [TRACE] EvalReadDiff: Read Create change from plan for module.roleDefinition.azurerm_role_definition.roleDefinition[0]
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalIf
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: terraform.EvalNoop
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalIf
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalReadState
2019/09/06 15:40:06 [TRACE] EvalReadState: reading state for module.roleDefinition.azurerm_role_definition.roleDefinition[0]
2019/09/06 15:40:06 [TRACE] EvalReadState: no state present for module.roleDefinition.azurerm_role_definition.roleDefinition[0]
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalReadDiff
2019/09/06 15:40:06 [TRACE] EvalReadDiff: Read Create change from plan for module.roleDefinition.azurerm_role_definition.roleDefinition[0]
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalDiff
2019/09/06 15:40:06 [TRACE] Re-validating config for "module.roleDefinition.azurerm_role_definition.roleDefinition[0]"
2019/09/06 15:40:06 [TRACE] GRPCProvider: ValidateResourceTypeConfig
2019/09/06 15:40:06 [TRACE] GRPCProvider: PlanResourceChange
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalCheckPlannedChange
2019/09/06 15:40:06 [TRACE] EvalCheckPlannedChange: Verifying that actual change (action Create) matches planned change (action Create)
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalGetProvider
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalReadState
2019/09/06 15:40:06 [TRACE] EvalReadState: reading state for module.roleDefinition.azurerm_role_definition.roleDefinition[0]
2019/09/06 15:40:06 [TRACE] EvalReadState: no state present for module.roleDefinition.azurerm_role_definition.roleDefinition[0]
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalReduceDiff
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalIf
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: terraform.EvalNoop
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalApplyPre
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalApply
2019/09/06 15:40:06 [DEBUG] module.roleDefinition.azurerm_role_definition.roleDefinition[0]: applying the planned Create change
2019/09/06 15:40:06 [TRACE] GRPCProvider: ApplyResourceChange
module.roleDefinition.azurerm_role_definition.roleDefinition[0]: Creating...
2019-09-06T15:40:06.492+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: 
Error: rpc error: code = Unavailable desc = transport is closing


Releasing state lock. This may take a few moments...
panic: interface conversion: interface {} is nil, not string
2019-09-06T15:40:06.492+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: 
2019-09-06T15:40:06.492+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: goroutine 59 [running]:
2019-09-06T15:40:06.492+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: github.com/terraform-providers/terraform-provider-azurerm/azurerm.expandRoleDefinitionPermissions(0xc0002ad260, 0xc0005e0ce0, 0xb, 0x3be7880)
2019-09-06T15:40:06.492+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/azurerm/resource_arm_role_definition.go:247 +0xab1
2019-09-06T15:40:06.492+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: github.com/terraform-providers/terraform-provider-azurerm/azurerm.resourceArmRoleDefinitionCreateUpdate(0xc0002ad260, 0x3dc13c0, 0xc000580500, 0x2, 0x7868aa0)
2019-09-06T15:40:06.492+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/azurerm/resource_arm_role_definition.go:118 +0x21a
2019-09-06T15:40:06.492+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/helper/schema.(*Resource).Apply(0xc000408480, 0xc000ce6320, 0xc0007c6b40, 0x3dc13c0, 0xc000580500, 0xc00065bb01, 0xc00097a098, 0x3c9c840)
2019-09-06T15:40:06.492+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/helper/schema/resource.go:287 +0x3b4
2019-09-06T15:40:06.492+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/helper/schema.(*Provider).Apply(0xc000742200, 0xc000a57a30, 0xc000ce6320, 0xc0007c6b40, 0xc000638868, 0xc0000b6f28, 0x3c9ed60)
2019-09-06T15:40:06.492+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/helper/schema/provider.go:285 +0x18f
2019-09-06T15:40:06.492+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/helper/plugin.(*GRPCProviderServer).ApplyResourceChange(0xc00000e9f0, 0x47ee220, 0xc000510000, 0xc00097c060, 0xc00000e9f0, 0xc000510000, 0xc0006c0ba8)
2019-09-06T15:40:06.492+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/helper/plugin/grpc_provider.go:885 +0x894
2019-09-06T15:40:06.492+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0x3fe47c0, 0xc00000e9f0, 0x47ee220, 0xc000510000, 0xc000ce6050, 0x0, 0x47ee220, 0xc000510000, 0xc000656000, 0x1b2c)
2019-09-06T15:40:06.492+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform/internal/tfplugin5/tfplugin5.pb.go:3217 +0x23e
2019-09-06T15:40:06.492+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc00008b200, 0x4839760, 0xc0000b4d80, 0xc00061e000, 0xc00056f1d0, 0x783ba00, 0x0, 0x0, 0x0)
2019-09-06T15:40:06.492+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:972 +0x470
2019-09-06T15:40:06.492+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc.(*Server).handleStream(0xc00008b200, 0x4839760, 0xc0000b4d80, 0xc00061e000, 0x0)
2019-09-06T15:40:06.492+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:1252 +0xda6
2019-09-06T15:40:06.492+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc000436b30, 0xc00008b200, 0x4839760, 0xc0000b4d80, 0xc00061e000)
2019-09-06T15:40:06.492+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:691 +0x9f
2019-09-06T15:40:06.493+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: created by github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
2019-09-06T15:40:06.493+0100 [DEBUG] plugin.terraform-provider-azurerm_v1.33.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:689 +0xa1
2019-09-06T15:40:06.497+0100 [DEBUG] plugin: plugin process exited: path=/Users/Lee/Code/terraform/Terraform.Environments/rbac/.terraform/plugins/darwin_amd64/terraform-provider-azurerm_v1.33.0_x4 pid=38358 error="exit status 2"
2019/09/06 15:40:06 [DEBUG] module.roleDefinition.azurerm_role_definition.roleDefinition[0]: apply errored, but we're indicating that via the Error pointer rather than returning it: rpc error: code = Unavailable desc = transport is closing
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalMaybeTainted
2019/09/06 15:40:06 [TRACE] EvalMaybeTainted: module.roleDefinition.azurerm_role_definition.roleDefinition[0] encountered an error during creation, so it is now marked as tainted
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalWriteState
2019/09/06 15:40:06 [TRACE] EvalWriteState: removing state object for module.roleDefinition.azurerm_role_definition.roleDefinition[0]
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalApplyProvisioners
2019/09/06 15:40:06 [TRACE] EvalApplyProvisioners: azurerm_role_definition.roleDefinition[0] has no state, so skipping provisioners
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalMaybeTainted
2019/09/06 15:40:06 [TRACE] EvalMaybeTainted: module.roleDefinition.azurerm_role_definition.roleDefinition[0] encountered an error during creation, so it is now marked as tainted
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalWriteState
2019/09/06 15:40:06 [TRACE] EvalWriteState: removing state object for module.roleDefinition.azurerm_role_definition.roleDefinition[0]
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalIf
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalIf
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalWriteDiff
2019/09/06 15:40:06 [TRACE] module.roleDefinition: eval: *terraform.EvalApplyPost
2019/09/06 15:40:06 [ERROR] module.roleDefinition: eval: *terraform.EvalApplyPost, err: rpc error: code = Unavailable desc = transport is closing
2019/09/06 15:40:06 [ERROR] module.roleDefinition: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing
2019/09/06 15:40:06 [TRACE] [walkApply] Exiting eval tree: module.roleDefinition.azurerm_role_definition.roleDefinition[0]
2019/09/06 15:40:06 [TRACE] vertex "module.roleDefinition.azurerm_role_definition.roleDefinition[0]": visit complete
2019/09/06 15:40:06 [TRACE] dag/walk: upstream of "module.roleDefinition.output.roleDefinitionId" errored, so skipping
2019/09/06 15:40:06 [TRACE] dag/walk: upstream of "module.roleAssignment.var.roleDefinitionId" errored, so skipping
2019/09/06 15:40:06 [TRACE] dag/walk: upstream of "module.roleAssignment.azurerm_role_assignment.roleAssignment[0]" errored, so skipping
2019/09/06 15:40:06 [TRACE] dag/walk: upstream of "module.roleAssignment.output.roleAssignmentId" errored, so skipping
2019/09/06 15:40:06 [TRACE] dag/walk: upstream of "meta.count-boundary (EachMode fixup)" errored, so skipping
2019/09/06 15:40:06 [TRACE] dag/walk: upstream of "provider.azurerm (close)" errored, so skipping
2019/09/06 15:40:06 [TRACE] dag/walk: upstream of "root" errored, so skipping
2019-09-06T15:40:07.321+0100 [DEBUG] plugin: plugin exited
make: *** [apply] Error 1

```

--- a/azurerm/internal/services/authorization/azuresdkhacks/definitions.go
+++ b/azurerm/internal/services/authorization/azuresdkhacks/definitions.go
@@ -4,7 +4,7 @@ import (
 	"context"
 	"net/http"
 
-	"github.com/Azure/azure-sdk-for-go/services/preview/authorization/mgmt/2018-09-01-preview/authorization"
+	"github.com/Azure/azure-sdk-for-go/services/preview/authorization/mgmt/2020-04-01-preview/authorization"
 	"github.com/Azure/go-autorest/autorest"
 	"github.com/Azure/go-autorest/autorest/azure"
 )

--- a/azurerm/internal/services/authorization/azuresdkhacks/definitions.go
+++ b/azurerm/internal/services/authorization/azuresdkhacks/definitions.go
@@ -1,0 +1,154 @@
+package azuresdkhacks
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/Azure/azure-sdk-for-go/services/preview/authorization/mgmt/2018-09-01-preview/authorization"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/azure"
+)
+
+type RoleDefinitionsWorkaroundClient struct {
+	sdkClient *authorization.RoleDefinitionsClient
+}
+
+func NewRoleDefinitionsWorkaroundClient(client *authorization.RoleDefinitionsClient) RoleDefinitionsWorkaroundClient {
+	return RoleDefinitionsWorkaroundClient{
+		sdkClient: client,
+	}
+}
+
+// CreateOrUpdate creates or updates a role definition.
+// Parameters:
+// scope - the scope of the role definition.
+// roleDefinitionID - the ID of the role definition.
+// roleDefinition - the values for the role definition.
+func (client RoleDefinitionsWorkaroundClient) CreateOrUpdate(ctx context.Context, scope string, roleDefinitionID string, roleDefinition authorization.RoleDefinition) (result RoleDefinitionUpdateResponse, err error) {
+	req, err := client.sdkClient.CreateOrUpdatePreparer(ctx, scope, roleDefinitionID, roleDefinition)
+	if err != nil {
+		err = autorest.NewErrorWithError(err, "authorization.RoleDefinitionsClient", "CreateOrUpdate", nil, "Failure preparing request")
+		return
+	}
+
+	resp, err := client.sdkClient.CreateOrUpdateSender(req)
+	if err != nil {
+		result.Response = autorest.Response{Response: resp}
+		err = autorest.NewErrorWithError(err, "authorization.RoleDefinitionsClient", "CreateOrUpdate", resp, "Failure sending request")
+		return
+	}
+
+	result, err = client.CreateOrUpdateResponder(resp)
+	if err != nil {
+		err = autorest.NewErrorWithError(err, "authorization.RoleDefinitionsClient", "CreateOrUpdate", resp, "Failure responding to request")
+	}
+
+	return
+}
+
+// CreateOrUpdateResponder handles the response to the CreateOrUpdate request. The method always
+// closes the http.Response Body.
+func (client RoleDefinitionsWorkaroundClient) CreateOrUpdateResponder(resp *http.Response) (result RoleDefinitionUpdateResponse, err error) {
+	err = autorest.Respond(
+		resp,
+		azure.WithErrorUnlessStatusCode(http.StatusOK, http.StatusCreated),
+		autorest.ByUnmarshallingJSON(&result),
+		autorest.ByClosing())
+	result.Response = autorest.Response{Response: resp}
+	return
+}
+
+// Get get role definition by name (GUID).
+// Parameters:
+// scope - the scope of the role definition.
+// roleDefinitionID - the ID of the role definition.
+func (client RoleDefinitionsWorkaroundClient) Get(ctx context.Context, scope string, roleDefinitionID string) (result RoleDefinitionGetResponse, err error) {
+	req, err := client.sdkClient.GetPreparer(ctx, scope, roleDefinitionID)
+	if err != nil {
+		err = autorest.NewErrorWithError(err, "authorization.RoleDefinitionsClient", "Get", nil, "Failure preparing request")
+		return
+	}
+
+	resp, err := client.sdkClient.GetSender(req)
+	if err != nil {
+		result.Response = autorest.Response{Response: resp}
+		err = autorest.NewErrorWithError(err, "authorization.RoleDefinitionsClient", "Get", resp, "Failure sending request")
+		return
+	}
+
+	result, err = client.GetResponder(resp)
+	if err != nil {
+		err = autorest.NewErrorWithError(err, "authorization.RoleDefinitionsClient", "Get", resp, "Failure responding to request")
+	}
+
+	return
+}
+
+// GetResponder handles the response to the Get request. The method always
+// closes the http.Response Body.
+func (client RoleDefinitionsWorkaroundClient) GetResponder(resp *http.Response) (result RoleDefinitionGetResponse, err error) {
+	err = autorest.Respond(
+		resp,
+		azure.WithErrorUnlessStatusCode(http.StatusOK),
+		autorest.ByUnmarshallingJSON(&result),
+		autorest.ByClosing())
+	result.Response = autorest.Response{Response: resp}
+	return
+}
+
+// RoleDefinition role definition.
+type RoleDefinitionGetResponse struct {
+	autorest.Response `json:"-"`
+	// ID - READ-ONLY; The role definition ID.
+	ID *string `json:"id,omitempty"`
+	// Name - READ-ONLY; The role definition name.
+	Name *string `json:"name,omitempty"`
+	// Type - READ-ONLY; The role definition type.
+	Type *string `json:"type,omitempty"`
+	// RoleDefinitionProperties - Role definition properties.
+	*RoleDefinitionProperties `json:"properties,omitempty"`
+}
+
+type RoleDefinitionUpdateResponse struct {
+	autorest.Response `json:"-"`
+	// ID - READ-ONLY; The role definition ID.
+	ID *string `json:"id,omitempty"`
+	// Name - READ-ONLY; The role definition name.
+	Name *string `json:"name,omitempty"`
+	// Type - READ-ONLY; The role definition type.
+	Type *string `json:"type,omitempty"`
+	// RoleDefinitionProperties - Role definition properties.
+	*RoleDefinitionProperties `json:"properties,omitempty"`
+}
+
+// RoleDefinitionProperties role definition properties.
+type RoleDefinitionProperties struct {
+	// RoleName - The role name.
+	RoleName *string `json:"roleName,omitempty"`
+	// Description - The role definition description.
+	Description *string `json:"description,omitempty"`
+	// RoleType - The role type.
+	RoleType *string `json:"type,omitempty"`
+	// Permissions - Role definition permissions.
+	Permissions *[]Permission `json:"permissions,omitempty"`
+	// AssignableScopes - Role definition assignable scopes.
+	AssignableScopes *[]string `json:"assignableScopes,omitempty"`
+
+	// not exposed in the sdk
+	CreatedOn *string `json:"createdOn,omitempty"`
+	UpdatedOn *string `json:"updatedOn,omitempty"`
+	CreatedBy *string `json:"createdBy,omitempty"`
+	UpdatedBy *string `json:"updatedBy,omitempty"`
+}
+
+// Permission role definition permissions.
+type Permission struct {
+	// Actions - Allowed actions.
+	Actions *[]string `json:"actions,omitempty"`
+	// NotActions - Denied actions.
+	NotActions *[]string `json:"notActions,omitempty"`
+	// DataActions - Allowed Data actions.
+	DataActions *[]string `json:"dataActions,omitempty"`
+	// NotDataActions - Denied Data actions.
+	NotDataActions *[]string `json:"notDataActions,omitempty"`
+}

--- a/azurerm/internal/services/authorization/role_definition_resource.go
+++ b/azurerm/internal/services/authorization/role_definition_resource.go
@@ -12,6 +12,7 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/clients"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/authorization/azuresdkhacks"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/authorization/parse"
 	azSchema "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/tf/schema"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/timeouts"
@@ -20,9 +21,9 @@ import (
 
 func resourceArmRoleDefinition() *schema.Resource {
 	return &schema.Resource{
-		Create: resourceArmRoleDefinitionCreateUpdate,
+		Create: resourceArmRoleDefinitionCreate,
 		Read:   resourceArmRoleDefinitionRead,
-		Update: resourceArmRoleDefinitionCreateUpdate,
+		Update: resourceArmRoleDefinitionUpdate,
 		Delete: resourceArmRoleDefinitionDelete,
 
 		Importer: azSchema.ValidateResourceIDPriorToImport(func(id string) error {
@@ -127,9 +128,9 @@ func resourceArmRoleDefinition() *schema.Resource {
 	}
 }
 
-func resourceArmRoleDefinitionCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+func resourceArmRoleDefinitionCreate(d *schema.ResourceData, meta interface{}) error {
 	client := meta.(*clients.Client).Authorization.RoleDefinitionsClient
-	ctx, cancel := timeouts.ForCreateUpdate(meta.(*clients.Client).StopContext, d)
+	ctx, cancel := timeouts.ForCreate(meta.(*clients.Client).StopContext, d)
 	defer cancel()
 
 	roleDefinitionId := d.Get("role_definition_id").(string)
@@ -215,6 +216,68 @@ func resourceArmRoleDefinitionCreateUpdate(d *schema.ResourceData, meta interfac
 	return resourceArmRoleDefinitionRead(d, meta)
 }
 
+func resourceArmRoleDefinitionUpdate(d *schema.ResourceData, meta interface{}) error {
+	sdkClient := meta.(*clients.Client).Authorization.RoleDefinitionsClient
+	client := azuresdkhacks.NewRoleDefinitionsWorkaroundClient(sdkClient)
+	ctx, cancel := timeouts.ForUpdate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	roleDefinitionId, err := parse.RoleDefinitionId(d.Id())
+	if err != nil {
+		return err
+	}
+
+	name := d.Get("name").(string)
+	description := d.Get("description").(string)
+	roleType := "CustomRole"
+
+	permissionsRaw := d.Get("permissions").([]interface{})
+	permissions := expandRoleDefinitionPermissions(permissionsRaw)
+	assignableScopes := expandRoleDefinitionAssignableScopes(d)
+
+	properties := authorization.RoleDefinition{
+		RoleDefinitionProperties: &authorization.RoleDefinitionProperties{
+			RoleName:         utils.String(name),
+			Description:      utils.String(description),
+			RoleType:         utils.String(roleType),
+			Permissions:      &permissions,
+			AssignableScopes: &assignableScopes,
+		},
+	}
+
+	resp, err := client.CreateOrUpdate(ctx, roleDefinitionId.Scope, roleDefinitionId.RoleID, properties)
+	if err != nil {
+		return fmt.Errorf("updating Role Definition %q (Scope %q): %+v", roleDefinitionId.RoleID, roleDefinitionId.Scope, err)
+	}
+	if resp.RoleDefinitionProperties == nil {
+		return fmt.Errorf("updating Role Definition %q (Scope %q): `properties` was nil", roleDefinitionId.RoleID, roleDefinitionId.Scope)
+	}
+	updatedOn := resp.RoleDefinitionProperties.UpdatedOn
+	if updatedOn == nil {
+		return fmt.Errorf("updating Role Definition %q (Scope %q): `properties.UpdatedOn` was nil", roleDefinitionId.RoleID, roleDefinitionId.Scope)
+	}
+
+	// "Updating" a role definition actually creates a new one and these get consolidated a few seconds later
+	// where the "create date" and "update date" match for the newly created record
+	// but eventually switch to being the old create date and the new update date
+	// ergo we can can for the old create date and the new updated date
+	log.Printf("[DEBUG] Waiting for Role Definition %q (Scope %q) to settle down..", roleDefinitionId.RoleID, roleDefinitionId.Scope)
+	stateConf := &resource.StateChangeConf{
+		ContinuousTargetOccurence: 5,
+		Delay:                     10 * time.Second,
+		MinTimeout:                10 * time.Second,
+		Pending:                   []string{"Pending"},
+		Target:                    []string{"Updated"},
+		Refresh:                   roleDefinitionEventualConsistencyUpdate(ctx, client, *roleDefinitionId, *updatedOn),
+		Timeout:                   d.Timeout(schema.TimeoutUpdate),
+	}
+	if _, err := stateConf.WaitForState(); err != nil {
+		return fmt.Errorf("waiting for Role Definition %q (Scope %q) to settle down: %+v", roleDefinitionId.RoleID, roleDefinitionId.Scope, err)
+	}
+
+	return resourceArmRoleDefinitionRead(d, meta)
+}
+
 func resourceArmRoleDefinitionRead(d *schema.ResourceData, meta interface{}) error {
 	client := meta.(*clients.Client).Authorization.RoleDefinitionsClient
 	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
@@ -293,6 +356,34 @@ func resourceArmRoleDefinitionDelete(d *schema.ResourceData, meta interface{}) e
 	return nil
 }
 
+func roleDefinitionEventualConsistencyUpdate(ctx context.Context, client azuresdkhacks.RoleDefinitionsWorkaroundClient, id parse.RoleDefinitionID, expectedUpdateDate string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		resp, err := client.Get(ctx, id.Scope, id.RoleID)
+		if err != nil {
+			return resp, "Failed", err
+		}
+		if resp.RoleDefinitionProperties == nil {
+			return resp, "Failed", fmt.Errorf("`properties` was nil")
+		}
+		if resp.RoleDefinitionProperties.CreatedOn == nil {
+			return resp, "Failed", fmt.Errorf("`properties.CreatedOn` was nil")
+		}
+
+		respCreatedOn := *resp.RoleDefinitionProperties.CreatedOn
+		respUpdatedOn := *resp.RoleDefinitionProperties.UpdatedOn
+		if respCreatedOn == expectedUpdateDate {
+			// a new role definition is created and eventually (~5s) reconciled
+			return resp, "Pending", nil
+		}
+		if respUpdatedOn != expectedUpdateDate {
+			// however the updatedOn should match the new date, to show this has been reconciled
+			return resp, "Pending", nil
+		}
+
+		return resp, "Updated", nil
+	}
+}
+
 func expandRoleDefinitionPermissions(input []interface{}) []authorization.Permission {
 	output := make([]authorization.Permission, 0)
 	if len(input) == 0 {
@@ -304,32 +395,32 @@ func expandRoleDefinitionPermissions(input []interface{}) []authorization.Permis
 			continue
 		}
 
-		input := v.(map[string]interface{})
+		raw := v.(map[string]interface{})
 		permission := authorization.Permission{}
 
 		actionsOutput := make([]string, 0)
-		actions := input["actions"].([]interface{})
+		actions := raw["actions"].([]interface{})
 		for _, a := range actions {
 			actionsOutput = append(actionsOutput, a.(string))
 		}
 		permission.Actions = &actionsOutput
 
 		dataActionsOutput := make([]string, 0)
-		dataActions := input["data_actions"].(*schema.Set)
+		dataActions := raw["data_actions"].(*schema.Set)
 		for _, a := range dataActions.List() {
 			dataActionsOutput = append(dataActionsOutput, a.(string))
 		}
 		permission.DataActions = &dataActionsOutput
 
 		notActionsOutput := make([]string, 0)
-		notActions := input["not_actions"].([]interface{})
+		notActions := raw["not_actions"].([]interface{})
 		for _, a := range notActions {
 			notActionsOutput = append(notActionsOutput, a.(string))
 		}
 		permission.NotActions = &notActionsOutput
 
 		notDataActionsOutput := make([]string, 0)
-		notDataActions := input["not_data_actions"].(*schema.Set)
+		notDataActions := raw["not_data_actions"].(*schema.Set)
 		for _, a := range notDataActions.List() {
 			notDataActionsOutput = append(notDataActionsOutput, a.(string))
 		}
@@ -366,9 +457,9 @@ func flattenRoleDefinitionPermissions(input *[]authorization.Permission) []inter
 	for _, permission := range *input {
 		permissions = append(permissions, map[string]interface{}{
 			"actions":          utils.FlattenStringSlice(permission.Actions),
-			"data_actions":     utils.FlattenStringSlice(permission.DataActions),
+			"data_actions":     schema.NewSet(schema.HashString, utils.FlattenStringSlice(permission.DataActions)),
 			"not_actions":      utils.FlattenStringSlice(permission.NotActions),
-			"not_data_actions": utils.FlattenStringSlice(permission.NotDataActions),
+			"not_data_actions": schema.NewSet(schema.HashString, utils.FlattenStringSlice(permission.NotDataActions)),
 		})
 	}
 

--- a/azurerm/internal/services/authorization/role_definition_resource.go
+++ b/azurerm/internal/services/authorization/role_definition_resource.go
@@ -73,7 +73,7 @@ func resourceArmRoleDefinition() *schema.Resource {
 
 			"permissions": {
 				Type:     schema.TypeList,
-				Required: true,
+				Optional: true,
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
 						"actions": {
@@ -146,7 +146,9 @@ func resourceArmRoleDefinitionCreateUpdate(d *schema.ResourceData, meta interfac
 	scope := d.Get("scope").(string)
 	description := d.Get("description").(string)
 	roleType := "CustomRole"
-	permissions := expandRoleDefinitionPermissions(d)
+
+	permissionsRaw := d.Get("permissions").([]interface{})
+	permissions := expandRoleDefinitionPermissions(permissionsRaw)
 	assignableScopes := expandRoleDefinitionAssignableScopes(d)
 
 	if d.IsNewResource() {
@@ -291,11 +293,13 @@ func resourceArmRoleDefinitionDelete(d *schema.ResourceData, meta interface{}) e
 	return nil
 }
 
-func expandRoleDefinitionPermissions(d *schema.ResourceData) []authorization.Permission {
+func expandRoleDefinitionPermissions(input []interface{}) []authorization.Permission {
 	output := make([]authorization.Permission, 0)
+	if len(input) == 0 {
+		return output
+	}
 
-	permissions := d.Get("permissions").([]interface{})
-	for _, v := range permissions {
+	for _, v := range input {
 		input := v.(map[string]interface{})
 		permission := authorization.Permission{}
 
@@ -356,37 +360,12 @@ func flattenRoleDefinitionPermissions(input *[]authorization.Permission) []inter
 	}
 
 	for _, permission := range *input {
-		output := make(map[string]interface{})
-
-		actions := make([]string, 0)
-		if s := permission.Actions; s != nil {
-			actions = *s
-		}
-		output["actions"] = actions
-
-		dataActions := make([]interface{}, 0)
-		if permission.DataActions != nil {
-			for _, dataAction := range *permission.DataActions {
-				dataActions = append(dataActions, dataAction)
-			}
-		}
-		output["data_actions"] = schema.NewSet(schema.HashString, dataActions)
-
-		notActions := make([]string, 0)
-		if s := permission.NotActions; s != nil {
-			notActions = *s
-		}
-		output["not_actions"] = notActions
-
-		notDataActions := make([]interface{}, 0)
-		if permission.NotDataActions != nil {
-			for _, dataAction := range *permission.NotDataActions {
-				notDataActions = append(notDataActions, dataAction)
-			}
-		}
-		output["not_data_actions"] = schema.NewSet(schema.HashString, notDataActions)
-
-		permissions = append(permissions, output)
+		permissions = append(permissions, map[string]interface{}{
+			"actions":          utils.FlattenStringSlice(permission.Actions),
+			"data_actions":     utils.FlattenStringSlice(permission.DataActions),
+			"not_actions":      utils.FlattenStringSlice(permission.NotActions),
+			"not_data_actions": utils.FlattenStringSlice(permission.NotDataActions),
+		})
 	}
 
 	return permissions

--- a/azurerm/internal/services/authorization/role_definition_resource.go
+++ b/azurerm/internal/services/authorization/role_definition_resource.go
@@ -356,7 +356,7 @@ func resourceArmRoleDefinitionDelete(d *schema.ResourceData, meta interface{}) e
 	return nil
 }
 
-func roleDefinitionEventualConsistencyUpdate(ctx context.Context, client azuresdkhacks.RoleDefinitionsWorkaroundClient, id parse.RoleDefinitionID, expectedUpdateDate string) resource.StateRefreshFunc {
+func roleDefinitionEventualConsistencyUpdate(ctx context.Context, client azuresdkhacks.RoleDefinitionsWorkaroundClient, id parse.RoleDefinitionID, updateRequestDate string) resource.StateRefreshFunc {
 	return func() (interface{}, string, error) {
 		resp, err := client.Get(ctx, id.Scope, id.RoleID)
 		if err != nil {
@@ -369,14 +369,19 @@ func roleDefinitionEventualConsistencyUpdate(ctx context.Context, client azuresd
 			return resp, "Failed", fmt.Errorf("`properties.CreatedOn` was nil")
 		}
 
-		respCreatedOn := *resp.RoleDefinitionProperties.CreatedOn
-		respUpdatedOn := *resp.RoleDefinitionProperties.UpdatedOn
-		if respCreatedOn == expectedUpdateDate {
+		if resp.RoleDefinitionProperties.UpdatedOn == nil {
+			return resp, "Failed", fmt.Errorf("`properties.UpdatedOn` was nil")
+		}
+
+		updateRequestTime, err := time.Parse(time.RFC3339, updateRequestDate)
+		respCreatedOn, _ := time.Parse(time.RFC3339, *resp.RoleDefinitionProperties.CreatedOn)
+		respUpdatedOn, _ := time.Parse(time.RFC3339, *resp.RoleDefinitionProperties.UpdatedOn)
+		if respCreatedOn.Equal(updateRequestTime) {
 			// a new role definition is created and eventually (~5s) reconciled
 			return resp, "Pending", nil
 		}
-		if respUpdatedOn != expectedUpdateDate {
-			// however the updatedOn should match the new date, to show this has been reconciled
+		if respUpdatedOn.Before(updateRequestTime) {
+			// The real updated on will the same as or after the time we requested it due to the swapout.
 			return resp, "Pending", nil
 		}
 

--- a/azurerm/internal/services/authorization/role_definition_resource.go
+++ b/azurerm/internal/services/authorization/role_definition_resource.go
@@ -374,8 +374,20 @@ func roleDefinitionEventualConsistencyUpdate(ctx context.Context, client azuresd
 		}
 
 		updateRequestTime, err := time.Parse(time.RFC3339, updateRequestDate)
-		respCreatedOn, _ := time.Parse(time.RFC3339, *resp.RoleDefinitionProperties.CreatedOn)
-		respUpdatedOn, _ := time.Parse(time.RFC3339, *resp.RoleDefinitionProperties.UpdatedOn)
+		if err != nil {
+			return nil, "", fmt.Errorf("error parsing time from update request: %+v", err)
+		}
+
+		respCreatedOn, err := time.Parse(time.RFC3339, *resp.RoleDefinitionProperties.CreatedOn)
+		if err != nil {
+			return nil, "", fmt.Errorf("error parsing time for createdOn from update request: %+v", err)
+		}
+
+		respUpdatedOn, err := time.Parse(time.RFC3339, *resp.RoleDefinitionProperties.UpdatedOn)
+		if err != nil {
+			return nil, "", fmt.Errorf("error parsing time for updatedOn from update request: %+v", err)
+		}
+
 		if respCreatedOn.Equal(updateRequestTime) {
 			// a new role definition is created and eventually (~5s) reconciled
 			return resp, "Pending", nil

--- a/azurerm/internal/services/authorization/role_definition_resource.go
+++ b/azurerm/internal/services/authorization/role_definition_resource.go
@@ -300,6 +300,10 @@ func expandRoleDefinitionPermissions(input []interface{}) []authorization.Permis
 	}
 
 	for _, v := range input {
+		if v == nil {
+			continue
+		}
+
 		input := v.(map[string]interface{})
 		permission := authorization.Permission{}
 

--- a/azurerm/internal/services/authorization/role_definition_resource.go
+++ b/azurerm/internal/services/authorization/role_definition_resource.go
@@ -401,6 +401,9 @@ func expandRoleDefinitionPermissions(input []interface{}) []authorization.Permis
 		actionsOutput := make([]string, 0)
 		actions := raw["actions"].([]interface{})
 		for _, a := range actions {
+			if a == nil {
+				continue
+			}
 			actionsOutput = append(actionsOutput, a.(string))
 		}
 		permission.Actions = &actionsOutput
@@ -408,6 +411,9 @@ func expandRoleDefinitionPermissions(input []interface{}) []authorization.Permis
 		dataActionsOutput := make([]string, 0)
 		dataActions := raw["data_actions"].(*schema.Set)
 		for _, a := range dataActions.List() {
+			if a == nil {
+				continue
+			}
 			dataActionsOutput = append(dataActionsOutput, a.(string))
 		}
 		permission.DataActions = &dataActionsOutput
@@ -415,6 +421,9 @@ func expandRoleDefinitionPermissions(input []interface{}) []authorization.Permis
 		notActionsOutput := make([]string, 0)
 		notActions := raw["not_actions"].([]interface{})
 		for _, a := range notActions {
+			if a == nil {
+				continue
+			}
 			notActionsOutput = append(notActionsOutput, a.(string))
 		}
 		permission.NotActions = &notActionsOutput
@@ -422,6 +431,9 @@ func expandRoleDefinitionPermissions(input []interface{}) []authorization.Permis
 		notDataActionsOutput := make([]string, 0)
 		notDataActions := raw["not_data_actions"].(*schema.Set)
 		for _, a := range notDataActions.List() {
+			if a == nil {
+				continue
+			}
 			notDataActionsOutput = append(notDataActionsOutput, a.(string))
 		}
 		permission.NotDataActions = &notDataActionsOutput

--- a/azurerm/internal/services/authorization/role_definition_resource.go
+++ b/azurerm/internal/services/authorization/role_definition_resource.go
@@ -34,7 +34,7 @@ func resourceArmRoleDefinition() *schema.Resource {
 		Timeouts: &schema.ResourceTimeout{
 			Create: schema.DefaultTimeout(30 * time.Minute),
 			Read:   schema.DefaultTimeout(5 * time.Minute),
-			Update: schema.DefaultTimeout(30 * time.Minute),
+			Update: schema.DefaultTimeout(60 * time.Minute),
 			Delete: schema.DefaultTimeout(30 * time.Minute),
 		},
 
@@ -263,8 +263,8 @@ func resourceArmRoleDefinitionUpdate(d *schema.ResourceData, meta interface{}) e
 	// ergo we can can for the old create date and the new updated date
 	log.Printf("[DEBUG] Waiting for Role Definition %q (Scope %q) to settle down..", roleDefinitionId.RoleID, roleDefinitionId.Scope)
 	stateConf := &resource.StateChangeConf{
-		ContinuousTargetOccurence: 5,
-		Delay:                     10 * time.Second,
+		ContinuousTargetOccurence: 12,
+		Delay:                     60 * time.Second,
 		MinTimeout:                10 * time.Second,
 		Pending:                   []string{"Pending"},
 		Target:                    []string{"Updated"},
@@ -392,8 +392,9 @@ func roleDefinitionEventualConsistencyUpdate(ctx context.Context, client azuresd
 			// a new role definition is created and eventually (~5s) reconciled
 			return resp, "Pending", nil
 		}
-		if respUpdatedOn.Before(updateRequestTime) {
-			// The real updated on will the same as or after the time we requested it due to the swapout.
+
+		if !respUpdatedOn.After(updateRequestTime) {
+			// The real updated on will be after the time we requested it due to the swap out.
 			return resp, "Pending", nil
 		}
 

--- a/azurerm/internal/services/authorization/role_definition_resource_test.go
+++ b/azurerm/internal/services/authorization/role_definition_resource_test.go
@@ -126,7 +126,7 @@ func TestAccRoleDefinition_emptyName(t *testing.T) {
 	})
 }
 
-func testAccRoleDefinition_emptyPermissions(t *testing.T) {
+func TestAccRoleDefinition_emptyPermissions(t *testing.T) {
 	data := acceptance.BuildTestData(t, "azurerm_role_definition", "test")
 	r := RoleDefinitionResource{}
 
@@ -141,7 +141,7 @@ func testAccRoleDefinition_emptyPermissions(t *testing.T) {
 	})
 }
 
-func testAccRoleDefinition_managementGroup(t *testing.T) {
+func TestAccRoleDefinition_managementGroup(t *testing.T) {
 	data := acceptance.BuildTestData(t, "azurerm_role_definition", "test")
 	r := RoleDefinitionResource{}
 

--- a/azurerm/internal/services/authorization/role_definition_resource_test.go
+++ b/azurerm/internal/services/authorization/role_definition_resource_test.go
@@ -16,7 +16,7 @@ import (
 
 type RoleDefinitionResource struct{}
 
-func TestAccAzureRMRoleDefinition_basic(t *testing.T) {
+func TestAccRoleDefinition_basic(t *testing.T) {
 	data := acceptance.BuildTestData(t, "azurerm_role_definition", "test")
 	r := RoleDefinitionResource{}
 
@@ -31,7 +31,7 @@ func TestAccAzureRMRoleDefinition_basic(t *testing.T) {
 	})
 }
 
-func TestAccAzureRMRoleDefinition_requiresImport(t *testing.T) {
+func TestAccRoleDefinition_requiresImport(t *testing.T) {
 	data := acceptance.BuildTestData(t, "azurerm_role_definition", "test")
 	id := uuid.New().String()
 
@@ -50,7 +50,7 @@ func TestAccAzureRMRoleDefinition_requiresImport(t *testing.T) {
 	})
 }
 
-func TestAccAzureRMRoleDefinition_complete(t *testing.T) {
+func TestAccRoleDefinition_complete(t *testing.T) {
 	data := acceptance.BuildTestData(t, "azurerm_role_definition", "test")
 	r := RoleDefinitionResource{}
 
@@ -65,7 +65,7 @@ func TestAccAzureRMRoleDefinition_complete(t *testing.T) {
 	})
 }
 
-func TestAccAzureRMRoleDefinition_update(t *testing.T) {
+func TestAccRoleDefinition_update(t *testing.T) {
 	data := acceptance.BuildTestData(t, "azurerm_role_definition", "test")
 	r := RoleDefinitionResource{}
 	id := uuid.New().String()
@@ -88,7 +88,7 @@ func TestAccAzureRMRoleDefinition_update(t *testing.T) {
 	})
 }
 
-func TestAccAzureRMRoleDefinition_updateEmptyId(t *testing.T) {
+func TestAccRoleDefinition_updateEmptyId(t *testing.T) {
 	data := acceptance.BuildTestData(t, "azurerm_role_definition", "test")
 
 	r := RoleDefinitionResource{}
@@ -111,7 +111,7 @@ func TestAccAzureRMRoleDefinition_updateEmptyId(t *testing.T) {
 	})
 }
 
-func TestAccAzureRMRoleDefinition_emptyName(t *testing.T) {
+func TestAccRoleDefinition_emptyName(t *testing.T) {
 	data := acceptance.BuildTestData(t, "azurerm_role_definition", "test")
 	r := RoleDefinitionResource{}
 
@@ -126,7 +126,22 @@ func TestAccAzureRMRoleDefinition_emptyName(t *testing.T) {
 	})
 }
 
-func TestAccAzureRMRoleDefinition_managementGroup(t *testing.T) {
+func TestAccRoleDefinition_emptyPermissions(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_role_definition", "test")
+	r := RoleDefinitionResource{}
+
+	data.ResourceTest(t, r, []resource.TestStep{
+		{
+			Config: r.emptyPermissions(data),
+			Check: resource.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.ImportStep(),
+	})
+}
+
+func TestAccRoleDefinition_managementGroup(t *testing.T) {
 	data := acceptance.BuildTestData(t, "azurerm_role_definition", "test")
 	r := RoleDefinitionResource{}
 
@@ -141,7 +156,7 @@ func TestAccAzureRMRoleDefinition_managementGroup(t *testing.T) {
 	})
 }
 
-func TestAccAzureRMRoleDefinition_assignToSmallerScope(t *testing.T) {
+func TestAccRoleDefinition_assignToSmallerScope(t *testing.T) {
 	data := acceptance.BuildTestData(t, "azurerm_role_definition", "test")
 	r := RoleDefinitionResource{}
 
@@ -156,7 +171,7 @@ func TestAccAzureRMRoleDefinition_assignToSmallerScope(t *testing.T) {
 	})
 }
 
-func TestAccAzureRMRoleDefinition_noAssignableScope(t *testing.T) {
+func TestAccRoleDefinition_noAssignableScope(t *testing.T) {
 	data := acceptance.BuildTestData(t, "azurerm_role_definition", "test")
 
 	r := RoleDefinitionResource{}
@@ -339,6 +354,28 @@ resource "azurerm_role_definition" "test" {
   ]
 }
 `, data.RandomInteger)
+}
+
+func (r RoleDefinitionResource) emptyPermissions(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+provider "azurerm" {
+  features {}
+}
+
+data "azurerm_subscription" "primary" {
+}
+
+resource "azurerm_resource_group" "test" {
+  name     = "acctestrg-%d"
+  location = %q
+}
+
+resource "azurerm_role_definition" "test" {
+  name              = "acctestrd-%d"
+  scope             = azurerm_resource_group.test.id
+  assignable_scopes = [azurerm_resource_group.test.id]
+}
+`, data.RandomInteger, data.Locations.Primary, data.RandomInteger)
 }
 
 func (RoleDefinitionResource) managementGroup(id string, data acceptance.TestData) string {

--- a/azurerm/internal/services/authorization/role_definition_resource_test.go
+++ b/azurerm/internal/services/authorization/role_definition_resource_test.go
@@ -126,7 +126,7 @@ func TestAccRoleDefinition_emptyName(t *testing.T) {
 	})
 }
 
-func TestAccRoleDefinition_emptyPermissions(t *testing.T) {
+func testAccRoleDefinition_emptyPermissions(t *testing.T) {
 	data := acceptance.BuildTestData(t, "azurerm_role_definition", "test")
 	r := RoleDefinitionResource{}
 
@@ -141,7 +141,7 @@ func TestAccRoleDefinition_emptyPermissions(t *testing.T) {
 	})
 }
 
-func TestAccRoleDefinition_managementGroup(t *testing.T) {
+func testAccRoleDefinition_managementGroup(t *testing.T) {
 	data := acceptance.BuildTestData(t, "azurerm_role_definition", "test")
 	r := RoleDefinitionResource{}
 

--- a/azurerm/internal/services/authorization/role_definition_resource_test.go
+++ b/azurerm/internal/services/authorization/role_definition_resource_test.go
@@ -292,7 +292,7 @@ resource "azurerm_role_definition" "test" {
   role_definition_id = "%s"
   name               = "acctestrd-%d"
   scope              = data.azurerm_subscription.primary.id
-  description        = "Acceptance Test Role Definition"
+  description        = "Acceptance Test Role Definition Updated"
 
   permissions {
     actions     = ["*"]

--- a/website/docs/r/role_definition.html.markdown
+++ b/website/docs/r/role_definition.html.markdown
@@ -45,11 +45,13 @@ The following arguments are supported:
 
 * `description` - (Optional) A description of the Role Definition.
 
-* `permissions` - (Required) A `permissions` block as defined below.
+* `permissions` - (Optional) A `permissions` block as defined below.
 
 * `assignable_scopes` - (Optional) One or more assignable scopes for this Role Definition, such as `/subscriptions/0b1f6471-1bf0-4dda-aec3-111122223333`, `/subscriptions/0b1f6471-1bf0-4dda-aec3-111122223333/resourceGroups/myGroup`, or `/subscriptions/0b1f6471-1bf0-4dda-aec3-111122223333/resourceGroups/myGroup/providers/Microsoft.Compute/virtualMachines/myVM`.
 
 ~> **NOTE:** The value for `scope` is automatically included in this list if no other values supplied.
+
+---
 
 A `permissions` block as the following properties:
 

